executor/src/witgen/jit/block_machine_processor.rs

```diff
@@ -0,0 +1,375 @@
+                    .map(move |id| (id, row))
+            })
+            .filter(|(identity, row)| !complete.contains(&(identity.id(), *row)))
+            .any(|(identity, _row)| !is_machine_call(identity));
+
+        match has_incomplete_machine_calls {
+            true => Err("Incomplete machine calls".to_string()),
+            false => Ok(()),
+        }
+    }
+}
+
+fn is_machine_call<T>(identity: &Identity<T>) -> bool {
+    matches!(
+        identity,
+        Identity::Lookup(_)
+            | Identity::Permutation(_)
+            | Identity::PhantomLookup(_)
+            | Identity::PhantomPermutation(_)
+            | Identity::PhantomBusInteraction(_)
+    )
+}
+
+pub struct FixedEvaluatorForFixedData<'a, T: FieldElement>(pub &'a FixedData<'a, T>);
+impl<'a, T: FieldElement> FixedEvaluator<T> for FixedEvaluatorForFixedData<'a, T> {
+    fn evaluate(&self, var: &AlgebraicReference, row_offset: i32) -> Option<T> {
+        assert!(var.is_fixed());
+        let values = self.0.fixed_cols[&var.poly_id].values_max_size();
+        let row = (row_offset + var.next as i32 + values.len() as i32) as usize % values.len();
+        Some(values[row])
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::collections::BTreeMap;
+
+    use bit_vec::BitVec;
+    use powdr_ast::analyzed::{
+        AlgebraicExpression, AlgebraicReference, Analyzed, SelectedExpressions,
+    };
+    use powdr_number::GoldilocksField;
+
+    use crate::{
+        constant_evaluator,
+        witgen::{
+            global_constraints,
+            jit::{affine_symbolic_expression::Effect, test_util::format_code},
+            machines::{Connection, ConnectionKind, MachineParts},
+            FixedData,
+        },
+    };
+
+    use super::{BlockMachineProcessor, Variable};
+
+    fn generate_for_block_machine(
+        input_pil: &str,
+        block_size: usize,
+        latch_row: usize,
+        selector_name: &str,
+        input_names: &[&str],
+        output_names: &[&str],
+    ) -> Result<Vec<Effect<GoldilocksField, Variable>>, String> {
+        let analyzed: Analyzed<GoldilocksField> =
+            powdr_pil_analyzer::analyze_string(input_pil).unwrap();
+        let fixed_col_vals = constant_evaluator::generate(&analyzed);
+        let fixed_data = FixedData::new(&analyzed, &fixed_col_vals, &[], Default::default(), 0);
+        let (fixed_data, retained_identities) =
+            global_constraints::set_global_constraints(fixed_data, &analyzed.identities);
+
+        // Build a connection that encodes:
+        // [] is <selector_name> $ [<input_names...>, <output_names...>]
+        let witnesses_by_name = analyzed
+            .committed_polys_in_source_order()
+            .flat_map(|(symbol, _)| symbol.array_elements())
+            .collect::<BTreeMap<_, _>>();
+        let to_expr = |name: &str| {
+            let next = name.ends_with("'");
+            let column_name = if next { &name[..name.len() - 1] } else { name };
```

Comment:
not sure if that is better, but it avoids the `&name[..name.len() - 1]` thing:
```suggestion
            let (column_name, next) = if let Some(column_name) = name.strip_suffix("'") { (column_name, true) } else { (name, false) };
```